File: Tyche.Backend/Pipel.Core/Pipel.Core/Page.fs
namespace Pipel.Core

type Page<'T> =
    { Items: 'T seq
      ItemsCount: int
      Skip: int
      Take: int
      HasNext: bool }

module Page =

    let map func page =
        { Page.Items = page.Items |> Seq.map func
          ItemsCount = page.ItemsCount
          Skip = page.Skip
          Take = page.Take
          HasNext = page.HasNext }

    let cast<'T> page =
        { Page.Items = page.Items |> Seq.cast<'T>
          ItemsCount = page.ItemsCount
          Skip = page.Skip
          Take = page.Take
          HasNext = page.HasNext }

    let fromSeq seq =
        let count = seq |> Seq.length

        { Page.Items = seq
          ItemsCount = count
          Skip = 0
          Take = count
          HasNext = false }

File: Tyche.Android/core/src/main/java/com/pipel/core/Page.kt
package com.pipel.core

data class Page<TModel : Any>(
    val items: List<TModel>,
    val itemsCount: Int,
    val skip: Int,
    val take: Int,
    val hasNext: Boolean,
) {

    fun <TOut : Any> map(mapFunc: MapFunc<TModel, TOut>): Page<TOut> {
        return Page(
            items = this.items.map(mapFunc),
            itemsCount = this.itemsCount,
            skip = this.skip,
            take = this.take,
            hasNext = this.hasNext
        )
    }

    companion object {
        fun <TModel : Any> empty(): Page<TModel> = Page(
            items = emptyList(),
            itemsCount = 0,
            skip = 0,
            take = 0,
            hasNext = false
        )
    }

}